src/natcap/invest/managed_raster/ManagedRaster.h

```diff
@@ -0,0 +1,607 @@
+#include <iostream>
+
+#include "gdal.h"
+#include "gdal_priv.h"
+#include <stdint.h>
+
+#include <errno.h>
+#include <string>
+#include <set>
+#include <cmath>
+#include <list>
+#include <utility>
+#include <iterator>
+
+#include "LRUCache.h"
+
+int MANAGED_RASTER_N_BLOCKS = pow(2, 6);
+// given the pixel neighbor numbering system
+//  3 2 1
+//  4 x 0
+//  5 6 7
+// These offsets are for the neighbor rows and columns
+int ROW_OFFSETS[8] = {0, -1, -1, -1,  0,  1, 1, 1};
+int COL_OFFSETS[8] = {1,  1,  0, -1, -1, -1, 0, 1};
+int FLOW_DIR_REVERSE_DIRECTION[8] = {4, 5, 6, 7, 0, 1, 2, 3};
+
+// if a pixel `x` has a neighbor `n` in position `i`,
+// then `n`'s neighbor in position `inflow_offsets[i]`
+// is the original pixel `x`
+int INFLOW_OFFSETS[8] = {4, 5, 6, 7, 0, 1, 2, 3};
+
+typedef std::pair<int, double*> BlockBufferPair;
+
+class NeighborTuple {
+public:
+    int direction, x, y;
+    float flow_proportion;
+
+    NeighborTuple () {}
+
+    NeighborTuple (int direction, int x, int y, float flow_proportion) {
+        this->direction = direction;
+        this->x = x;
+        this->y = y;
+        this->flow_proportion = flow_proportion;
+    }
+
+    ~NeighborTuple () {}
+};
+
+
+class ManagedRaster {
```

Comment:
I ran into issues with implementing a destructor because it seems to get called before I'd expect (such as when the cython object is created and assigned to a variable, it's actually copied and the original destroyed), possibly due to Cython weirdness? So I've put all the clean-up into `close` instead, so we can control when it's called. The default destructor should take care of everything else. (I might be missing something here, though!)